vms/evm/sync/customrawdb/markers.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	if len(upgrade) == 0 {
+		return config, nil
+	}
+
+	if err := json.Unmarshal(upgrade, &upgradeConfig); err != nil {
+		return nil, errInvalidData
+	}
+
+	return config, nil
+}
+
+// WriteChainConfig writes the chain config settings to the database.
+// The provided `upgradeConfig` (any JSON-marshalable type) will be stored alongside the chain config.
+func WriteChainConfig[T any](db ethdb.KeyValueWriter, hash common.Hash, config *params.ChainConfig, upgradeConfig T) error {
+	rawdb.WriteChainConfig(db, hash, config)
+	if config == nil {
+		return nil
+	}
+
+	data, err := json.Marshal(upgradeConfig)
+	if err != nil {
+		return fmt.Errorf("%w: %w", errFailedToMarshalUpgradeConfig, err)
+	}
+	if err := db.Put(upgradeConfigKey(hash), data); err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewAccountSnapshotsIterator returns an iterator for walking all of the accounts in the snapshot
+func NewAccountSnapshotsIterator(db ethdb.Iteratee) ethdb.Iterator {
+	it := db.NewIterator(rawdb.SnapshotAccountPrefix, nil)
+	keyLen := len(rawdb.SnapshotAccountPrefix) + common.HashLength
+	return rawdb.NewKeyLengthIterator(it, keyLen)
+}
+
+// ReadSnapshotBlockHash retrieves the hash of the block whose state is contained in
+// the persisted snapshot.
+func ReadSnapshotBlockHash(db ethdb.KeyValueReader) (common.Hash, error) {
+	ok, err := db.Has(snapshotBlockHashKey)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if !ok {
+		return common.Hash{}, ErrEntryNotFound
+	}
+
+	data, err := db.Get(snapshotBlockHashKey)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if len(data) != common.HashLength {
+		return common.Hash{}, fmt.Errorf("%w: length %d", errInvalidData, len(data))
+	}
+	return common.BytesToHash(data), nil
+}
+
+// WriteSnapshotBlockHash stores the root of the block whose state is contained in
+// the persisted snapshot.
+func WriteSnapshotBlockHash(db ethdb.KeyValueWriter, blockHash common.Hash) error {
+	return db.Put(snapshotBlockHashKey, blockHash[:])
+}
+
+// DeleteSnapshotBlockHash deletes the hash of the block whose state is contained in
+// the persisted snapshot. Since snapshots are not immutable, this  method can
+// be used during updates, so a crash or failure will mark the entire snapshot
+// invalid.
+func DeleteSnapshotBlockHash(db ethdb.KeyValueWriter) error {
+	return db.Delete(snapshotBlockHashKey)
+}
+
+func writeTimeMarker(db ethdb.KeyValueStore, key []byte, ts time.Time) error {
+	data, err := rlp.EncodeToBytes(uint64(ts.Unix()))
+	if err != nil {
+		return err
+	}
+	return db.Put(key, data)
+}
+
+func readTimeMarker(db ethdb.KeyValueStore, key []byte) (time.Time, error) {
+	// Check existence first to map missing marker to a stable sentinel error.
+	ok, err := db.Has(key)
+	if err != nil {
+		return time.Time{}, err
+	}
+	if !ok {
+		return time.Time{}, ErrEntryNotFound
+	}
+
+	data, err := db.Get(key)
+	if err != nil {
+		return time.Time{}, err
+	}
+	if len(data) == 0 {
```

Comment:
can this ever happen normally? i'm wondering f `errInvalidData` is a better error for this.

Reply:
Good catch - with our writers, the value is always RLP-encoded uint64 (non-empty). An empty value implies corruption or external misuse, not a normal state, so I'll change the error.